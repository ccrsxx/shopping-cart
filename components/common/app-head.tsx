import Head from 'next/head';

export function AppHead(): JSX.Element {
  return (
    <Head>
      <link rel='icon' href='/favicon.ico' />
      <link rel='manifest' href='/site.webmanifest' key='site-manifest' />
      <meta name='twitter:site' content='@ccrsxx' />
      <meta name='twitter:card' content='summary_large_image' />
      <meta name='og:title' content='Shopping Cart' />
      <meta name='og:url' content='https://shopping-cart-ccrsxx.vercel.app' />
      <meta
        name='description'
        content='The One Stop Shop for All Your Shopping Needs.'
      />
      <meta
        name='og:description'
        content='The One Stop Shop for All Your Shopping Needs.'
      />
      <meta
        property='og:image'
        content='https://og-image.vercel.app/Shopping%20Cart.png?theme=dark&md=1&fontSize=100px&images=https%3A%2F%2Fassets.vercel.com%2Fimage%2Fupload%2Ffront%2Fassets%2Fdesign%2Fvercel-triangle-white.svg'
      />
    </Head>
  );
}
