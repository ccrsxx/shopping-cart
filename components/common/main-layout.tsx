import Head from 'next/head';
import type { ReactNode } from 'react';

type MainLayoutProps = {
  className?: string;
  title: string;
  url?: string;
  description: string;
  image: string;
  children: ReactNode;
  product?: boolean;
};

export function MainLayout({
  className,
  title,
  url,
  description,
  image,
  children
}: MainLayoutProps): JSX.Element {
  const siteTitle = title;
  const siteDescription = description;
  const siteImage = image;

  return (
    <main className={className}>
      <Head>
        <title>{siteTitle}</title>
        <meta name='og:title' content={siteTitle} />
        <meta name='description' content={siteDescription} />
        <meta name='og:description' content={siteDescription} />
        <meta property='og:image' content={siteImage} />
        <meta
          name='og:url'
          content={`https://shopping-cart-ccrsxx.vercel.app${url ?? ''}`}
        />
      </Head>
      {children}
    </main>
  );
}
