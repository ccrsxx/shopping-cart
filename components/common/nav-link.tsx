import Link from 'next/link';
import cn from 'clsx';
import { delayScroll } from '@lib/scroll';

type NavLinkProps = {
  href: string;
  linkName: string;
  pathname: string;
  disableScroll?: boolean;
};

export function NavLink({
  href,
  linkName,
  pathname,
  disableScroll
}: NavLinkProps): JSX.Element {
  return (
    <Link href={href} key={href} scroll={!disableScroll}>
      <a
        className={cn(
          'tab hover:brightness-125',
          pathname === href && 'text-primary'
        )}
        onClick={disableScroll ? delayScroll : undefined}
      >
        {linkName}
      </a>
    </Link>
  );
}
