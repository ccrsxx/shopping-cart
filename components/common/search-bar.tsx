import { useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import cn from 'clsx';
import { useAnimation, motion } from 'framer-motion';
import { useShoppingCart } from '@lib/context/shopping-cart';
import { MdSearch } from '@assets/icons';
import { Button } from '@components/ui/button';
import type { ChangeEvent, FormEvent } from 'react';
import type { UrlObject } from 'url';
import type { QueryType } from '@components/store/aside';

export type ExtraQueryType = QueryType & {
  push: (url: UrlObject | string) => Promise<boolean>;
};

export function SearchBar(): JSX.Element {
  const { isMobile } = useShoppingCart();

  const {
    pathname,
    query: { search, category },
    push
  } = useRouter() as ExtraQueryType;

  const [searchInput, setSearchInput] = useState('');
  const [isExpanded, setIsExpanded] = useState(false);

  const controls = useAnimation();

  useEffect(() => {
    if (!search) setSearchInput('');
    else setSearchInput(search);
  }, [search, pathname]);

  useEffect(() => {
    setIsExpanded(isMobile);
  }, [isMobile]);

  const setMaxWidth = (maxWidth: number) => (): void =>
    void controls.start({
      maxWidth
    });

  const handleChange = ({
    target: { value }
  }: ChangeEvent<HTMLInputElement>): void => setSearchInput(value);

  const handleSubmit = (e: FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    void push({
      pathname: '/store',
      query: {
        ...(searchInput && { search: searchInput }),
        ...(category && { category })
      }
    });
  };

  return (
    <motion.form
      className={cn(
        `flex w-full flex-1 items-center gap-2 focus-within:max-w-sm
         md:flex-initial lg:-translate-x-[25%]`,
        isExpanded && '!max-w-full'
      )}
      initial={{ maxWidth: 300 }}
      animate={controls}
      onSubmit={handleSubmit}
    >
      <input
        type='text'
        className='tab w-full rounded-lg border border-neutral-500 bg-inherit px-2 py-1'
        placeholder='Search product'
        onFocus={setMaxWidth(500)}
        onBlur={setMaxWidth(300)}
        onChange={handleChange}
        value={searchInput}
      />
      <Button type='submit' className='!p-2 text-2xl' Icon={MdSearch} />
    </motion.form>
  );
}
