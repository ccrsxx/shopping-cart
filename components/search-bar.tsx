import { useState } from 'react';
import { useAnimation, motion } from 'framer-motion';
import { useWindowSize } from '@hooks/useWindowSize';
import { MdSearch } from '@components/icons';
import { Button } from '@components/button';

export function SearchBar(): JSX.Element {
  const [searchInput, setSearchInput] = useState('');
  const [width] = useWindowSize();

  const controls = useAnimation();

  // useEffect(() => {
  //   const searchParams = parameter.get('search');

  //   if (!searchParams) setSearchInput('');
  //   else setSearchInput(searchParams);
  // }, [pathname]);

  const setMaxWidth = (maxWidth: number) => (): void =>
    void controls.start({
      maxWidth
    });

  const handleChange = ({
    target: { value }
  }: React.ChangeEvent<HTMLInputElement>): void => setSearchInput(value);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>): void => {
    e.preventDefault();

    // const categoryParam = parameter.get('category');

    // const newParameter = new URLSearchParams({
    //   search: searchInput,
    //   ...(categoryParam && { category: categoryParam })
    // });

    // navigate(`/store?${newParameter}`);
  };

  const isMobile = width && width < 768;

  return (
    <motion.form
      className={`${
        isMobile ? '!max-w-full' : ''
      } flex w-full flex-1 items-center gap-2 focus-within:max-w-sm
        md:flex-initial lg:-translate-x-[25%]`}
      initial={{ maxWidth: 300 }}
      animate={controls}
      onSubmit={handleSubmit}
    >
      <input
        type='text'
        className='tab w-full rounded-lg border border-neutral-500 bg-inherit px-2 py-1'
        placeholder='Search product...'
        onFocus={setMaxWidth(500)}
        onBlur={setMaxWidth(300)}
        onChange={handleChange}
        value={searchInput}
      />
      <Button type='submit' className='!p-2 text-2xl' Icon={MdSearch} />
    </motion.form>
  );
}
