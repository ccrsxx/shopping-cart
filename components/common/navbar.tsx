import { useRouter } from 'next/router';
import { Cart } from '@components/cart/cart';
import { NavLink } from './nav-link';
import { SearchBar } from './search-bar';

const navLinks = [
  ['Home', '/'],
  ['Store', '/store'],
  ['Contact', '/contact']
];

export function Navbar(): JSX.Element {
  const { pathname } = useRouter();

  const inStore = pathname === '/store';

  return (
    <header
      className='fixed z-10 grid w-full gap-2 border-b-2 border-b-border-primary 
                 bg-background/75 backdrop-blur-md md:grid-cols-2'
    >
      <nav className='flex items-center justify-center gap-6 text-2xl text-secondary md:justify-start'>
        {navLinks.map(([linkName, url]) => (
          <NavLink
            href={url}
            linkName={linkName}
            pathname={pathname}
            disableScroll={inStore}
            key={url}
          />
        ))}
      </nav>
      <div className='flex justify-between gap-2 md:gap-0'>
        <SearchBar />
        <Cart />
      </div>
    </header>
  );
}
