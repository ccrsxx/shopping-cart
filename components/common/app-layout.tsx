import { Navbar } from './navbar';
import { Footer } from './footer';

type AppLayoutProps = {
  children: React.ReactNode;
};

export function AppLayout({ children }: AppLayoutProps): JSX.Element {
  return (
    <>
      <Navbar />
      {children}
      <Footer />
    </>
  );
}
