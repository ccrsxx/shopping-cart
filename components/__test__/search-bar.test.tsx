import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { SearchBar } from '@components/common/search-bar';
import type { ExtraQueryType } from '@components/common/search-bar';

const mockRouterPush = jest.fn();

jest.mock('next/router', () => ({
  useRouter: (): Omit<ExtraQueryType, 'isReady'> => ({
    pathname: 'home',
    query: { search: '', category: '' },
    push: mockRouterPush
  })
}));

jest.mock('@lib/context/shopping-cart', () => ({
  useShoppingCart: (): { isMobile: boolean } => ({ isMobile: false })
}));

describe('test SearchBar', () => {
  beforeEach(() => {
    render(<SearchBar />);
  });

  it('renders correctly', () => {
    const { container } = render(<SearchBar />);
    expect(container).toMatchSnapshot();
  });

  it('should have input value from search parameter', () => {
    const input = screen.getByRole('textbox');

    expect(input).toHaveValue('');
  });

  it('can receive input value', () => {
    const input = screen.getByRole('textbox');

    userEvent.type(input, 'test');

    expect(input).toHaveValue('test');
  });

  it('can submit the form and navigate to input value', () => {
    const input = screen.getByRole('textbox');

    userEvent.type(input, 'test');
    userEvent.type(input, '{enter}');

    expect(mockRouterPush).toHaveBeenCalledWith({
      pathname: '/store',
      query: {
        search: 'test'
      }
    });
  });

  it('can submit with search button', () => {
    const input = screen.getByRole('textbox');
    const button = screen.getByRole('button');

    userEvent.type(input, 'test');
    userEvent.click(button);

    expect(mockRouterPush).toHaveBeenCalledWith({
      pathname: '/store',
      query: {
        search: 'test'
      }
    });
  });

  it('can submit empty value', () => {
    const button = screen.getByRole('button');

    userEvent.click(button);

    expect(mockRouterPush).toHaveBeenCalledWith({
      pathname: '/store',
      query: {}
    });
  });
});
