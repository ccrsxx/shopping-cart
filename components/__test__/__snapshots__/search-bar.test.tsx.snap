// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test SearchBar renders correctly 1`] = `
<div>
  <form
    class="flex w-full flex-1 items-center gap-2 focus-within:max-w-sm
         md:flex-initial lg:-translate-x-[25%]"
    style="max-width: 300px;"
  >
    <input
      class="tab w-full rounded-lg border border-neutral-500 bg-inherit px-2 py-1"
      placeholder="Search product"
      type="text"
      value=""
    />
    <button
      class="tab flex select-none items-center justify-center gap-2 rounded-lg bg-transparent py-2 px-3 
         font-medium transition duration-300 hover:text-primary active:scale-90 active:duration-150 
         enabled:hover:bg-accent disabled:cursor-not-allowed disabled:brightness-50 !p-2 text-2xl"
      type="submit"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
        <path
          d="M15.5 14h-.79l-.28-.27A6.471 6.471 0 0016 9.5 6.5 6.5 0 109.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
        />
      </svg>
    </button>
  </form>
</div>
`;
